CMakeLists.txt

```diff
@@ -16,97 +16,115 @@ option(BUILD_TESTING "Enable test (depends on googletest)." OFF)
 # it prominent in the GUI.
 option(BUILD_SHARED_LIBS "Build library as shared." OFF)
 
-#
-# library : cpu_features
-#
-
 include (CheckIncludeFile)
 include (CheckSymbolExists)
 
-check_include_file(dlfcn.h HAVE_DLFCN_H)
-check_symbol_exists(getauxval "sys/auxv.h" HAVE_STRONG_GETAUXVAL)
+macro(setup_include_and_definitions TARGET_NAME)
+  target_include_directories(${TARGET_NAME}
+    PUBLIC  $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include>
+    PRIVATE $<BUILD_INTERFACE:${PROJECT_SOURCE_DIR}/include/internal>
+  )
+  target_compile_definitions(${TARGET_NAME}
+    PUBLIC STACK_LINE_READER_BUFFER_SIZE=1024
+  )
+endmacro()
 
-set(_HDRS
-  include/cpu_features_macros.h
-)
+set(PROCESSOR_IS_MIPS FALSE)
+set(PROCESSOR_IS_ARM FALSE)
+set(PROCESSOR_IS_AARCH64 FALSE)
+set(PROCESSOR_IS_X86 FALSE)
+set(PROCESSOR_IS_POWER FALSE)
+
+if(CMAKE_SYSTEM_PROCESSOR MATCHES "^mips")
+  set(PROCESSOR_IS_MIPS TRUE)
+endif()
+if(CMAKE_SYSTEM_PROCESSOR MATCHES "^arm")
+  set(PROCESSOR_IS_ARM TRUE)
+endif()
+if(CMAKE_SYSTEM_PROCESSOR MATCHES "^aarch64")
+  set(PROCESSOR_IS_AARCH64 TRUE)
+endif()
+if(CMAKE_SYSTEM_PROCESSOR MATCHES "(x86_64)|(AMD64)|(^i.86$)")
+  set(PROCESSOR_IS_X86 TRUE)
+endif()
+if(CMAKE_SYSTEM_PROCESSOR MATCHES "^(powerpc|ppc)")
+  set(PROCESSOR_IS_POWER TRUE)
+endif()
 
-set(_SRCS
+macro(add_arch_sources HDRS_LIST_NAME SRCS_LIST_NAME)
+  list(APPEND ${HDRS_LIST_NAME} include/cpu_features_macros.h)
+  if(PROCESSOR_IS_MIPS)
+      list(APPEND ${HDRS_LIST_NAME} include/cpuinfo_mips.h)
+      list(APPEND ${SRCS_LIST_NAME} src/cpuinfo_mips.c)
+  elseif(PROCESSOR_IS_ARM)
+      list(APPEND ${HDRS_LIST_NAME} include/cpuinfo_arm.h)
+      list(APPEND ${SRCS_LIST_NAME} src/cpuinfo_arm.c)
+  elseif(PROCESSOR_IS_AARCH64)
+      list(APPEND ${HDRS_LIST_NAME} include/cpuinfo_aarch64.h)
+      list(APPEND ${SRCS_LIST_NAME} src/cpuinfo_aarch64.c)
+  elseif(PROCESSOR_IS_X86)
+      list(APPEND ${HDRS_LIST_NAME} include/cpuinfo_x86.h)
+      list(APPEND ${HDRS_LIST_NAME} include/internal/cpuid_x86.h)
+      list(APPEND ${SRCS_LIST_NAME} src/cpuinfo_x86.c)
+  elseif(PROCESSOR_IS_POWER)
+      list(APPEND ${HDRS_LIST_NAME} include/cpuinfo_ppc.h)
+      list(APPEND ${SRCS_LIST_NAME} src/cpuinfo_ppc.c)
+  else()
+    message(FATAL_ERROR "Unsupported architectures ${CMAKE_SYSTEM_PROCESSOR}")
+  endif()
+endmacro()
+
+#
+# library : utils
+#
+
+add_library(utils OBJECT
   include/internal/bit_utils.h
+  include/internal/filesystem.h
+  include/internal/stack_line_reader.h
+  include/internal/string_view.h
+  src/filesystem.c
+  src/stack_line_reader.c
+  src/string_view.c
 )
+setup_include_and_definitions(utils)
 
-macro(add_linux_detection)
-  if(NOT UNIX)
-    message(FATAL_ERROR "Use of add_linux_detection() on non Linux-like OS")
-  endif()
-  list(APPEND _SRCS
-    include/internal/linux_features_aggregator.h
-    src/linux_features_aggregator.c
+#
+# library : linux_based_hardware_detection
+#
+
+if(UNIX)
+  add_library(linux_based_hardware_detection OBJECT
```

Comment:
`UNIX` include `APPLE` in CMake are you ok ?
should you rename linux_based_hardware... to unix_based_hardware... ?

Reply:
Yes thx I'll rename all the files as well